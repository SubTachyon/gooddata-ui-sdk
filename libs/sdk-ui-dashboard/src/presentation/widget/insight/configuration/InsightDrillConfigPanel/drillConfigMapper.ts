// (C) 2022 GoodData Corporation
import {
    IDrillToDashboard,
    IDrillToInsight,
    InsightDrillDefinition,
    isDrillFromAttribute,
    isDrillToAttributeUrl,
    isDrillToCustomUrl,
    isDrillToDashboard,
    isDrillToInsight,
    objRefToString,
} from "@gooddata/sdk-model";
import { IAvailableDrillTargets } from "@gooddata/sdk-ui";
import { defineMessage } from "react-intl";
import {
    getDrillOriginLocalIdentifier,
    getLocalIdentifierOrDie,
    getValidDrillOriginAttributes,
} from "../../../../../_staging/drills/drillingUtils.js";
import {
    DRILL_TARGET_TYPE,
    IDrillConfigItem,
    IDrillConfigItemBase,
    IDrillDownAttributeHierarchyConfig,
    IDrillToDashboardConfig,
    IDrillToInsightConfig,
    IDrillToUrl,
    IDrillToUrlConfig,
    isDrillToUrl,
    UrlDrillTarget,
} from "../../../../drill/types.js";
import { IGlobalDrillDownAttributeHierarchyDefinition } from "../../../../../types.js";

function getTitleFromDrillableItemPushData(items: IAvailableDrillTargets, itemId: string): string {
    const measureItems = items.measures || [];
    const measureResult = measureItems.find(
        (measureResult) => measureResult.measure.measureHeaderItem.localIdentifier === itemId,
    );

    if (measureResult) {
        return measureResult.measure.measureHeaderItem.name;
    }

    const attributeItems = items.attributes || [];
    const attributeResult = attributeItems.find(
        (attributeItem) => attributeItem.attribute.attributeHeader.localIdentifier === itemId,
    );

    return attributeResult ? attributeResult.attribute.attributeHeader.formOf.name : "";
}

const buildUrlDrillTarget = (drillData: IDrillToUrl): UrlDrillTarget => {
    if (isDrillToCustomUrl(drillData)) {
        const customUrl = drillData.target.url;
        return {
            customUrl,
        };
    }
    if (isDrillToAttributeUrl(drillData)) {
        const {
            displayForm: insightAttributeDisplayForm,
            hyperlinkDisplayForm: drillToAttributeDisplayForm,
        } = drillData.target;

        return {
            insightAttributeDisplayForm,
            drillToAttributeDisplayForm,
        };
    }
    throw new Error("Unsupported URL drill type!");
};

const createInsightConfig = (
    drillData: IDrillToInsight,
    supportedItemsForWidget: IAvailableDrillTargets,
): IDrillToInsightConfig => {
    const originLocalIdentifier = isDrillFromAttribute(drillData.origin)
        ? getLocalIdentifierOrDie(drillData.origin?.attribute)
        : getLocalIdentifierOrDie(drillData.origin?.measure);

    return {
        type: isDrillFromAttribute(drillData.origin) ? "attribute" : "measure",
        originLocalIdentifier,
        title: getTitleFromDrillableItemPushData(supportedItemsForWidget, originLocalIdentifier),
        attributes: getValidDrillOriginAttributes(supportedItemsForWidget, originLocalIdentifier),
        localIdentifier: drillData.localIdentifier!,
        drillTargetType: DRILL_TARGET_TYPE.DRILL_TO_INSIGHT,
        insightRef: drillData.target,
        complete: true,
    };
};

const createDashboardConfig = (
    drillData: IDrillToDashboard,
    supportedItemsForWidget: IAvailableDrillTargets,
): IDrillToDashboardConfig => {
    const originLocalIdentifier = isDrillFromAttribute(drillData.origin)
        ? getLocalIdentifierOrDie(drillData.origin?.attribute)
        : getLocalIdentifierOrDie(drillData.origin?.measure);

    return {
        type: isDrillFromAttribute(drillData.origin) ? "attribute" : "measure",
        originLocalIdentifier,
        title: getTitleFromDrillableItemPushData(supportedItemsForWidget, originLocalIdentifier),
        attributes: getValidDrillOriginAttributes(supportedItemsForWidget, originLocalIdentifier),
        localIdentifier: drillData.localIdentifier!,
        drillTargetType: DRILL_TARGET_TYPE.DRILL_TO_DASHBOARD,
        dashboard: drillData.target,
        complete: true,
    };
};

const invalidUrlMessage = defineMessage({
    id: "configurationPanel.drillConfig.drillIntoUrl.invalidCustomUrl",
});

const createUrlConfig = (
    drillData: IDrillToUrl,
    supportedItemsForWidget: IAvailableDrillTargets,
    invalidCustomUrlDrillLocalIds: string[],
): IDrillToUrlConfig => {
    const originLocalIdentifier = getDrillOriginLocalIdentifier(drillData);

    const hasWarning = invalidCustomUrlDrillLocalIds.includes(originLocalIdentifier);

    return {
        type: isDrillFromAttribute(drillData.origin) ? "attribute" : "measure",
        originLocalIdentifier,
        title: getTitleFromDrillableItemPushData(supportedItemsForWidget, originLocalIdentifier),
        attributes: getValidDrillOriginAttributes(supportedItemsForWidget, originLocalIdentifier),
        localIdentifier: drillData.localIdentifier!,
        drillTargetType: DRILL_TARGET_TYPE.DRILL_TO_URL,
        urlDrillTarget: buildUrlDrillTarget(drillData),
        complete: true,
        warning: hasWarning ? invalidUrlMessage.id : undefined,
    };
};

const createImplicitConfig = (): IDrillConfigItemBase => {
    return null as unknown as IDrillConfigItemBase;
};

const createConfig = (
    drillData: InsightDrillDefinition,
    supportedItemsForWidget: IAvailableDrillTargets,
    invalidCustomUrlDrillLocalIds: string[],
) => {
    if (isDrillToInsight(drillData)) {
        return createInsightConfig(drillData, supportedItemsForWidget);
    }

    if (isDrillToDashboard(drillData)) {
        return createDashboardConfig(drillData, supportedItemsForWidget);
    }

    if (isDrillToUrl(drillData)) {
        return createUrlConfig(drillData, supportedItemsForWidget, invalidCustomUrlDrillLocalIds);
    }

    return createImplicitConfig();
};

/**
 * @internal
 */
export const getMappedConfigForWidget = (
    configForWidget: InsightDrillDefinition[],
    supportedItemsForWidget: IAvailableDrillTargets,
    invalidCustomUrlDrillLocalIds: string[],
): IDrillConfigItem[] => {
    return configForWidget.map((item) =>
        createConfig(item, supportedItemsForWidget, invalidCustomUrlDrillLocalIds),
    );
};

/**
 * @internal
 */
export const getGlobalDrillDownMappedConfigForWidget = (
    widgetGlobalDrillDowns: IGlobalDrillDownAttributeHierarchyDefinition[],
    supportedItemsForWidget: IAvailableDrillTargets,
): IDrillDownAttributeHierarchyConfig[] => {
    return widgetGlobalDrillDowns.map((globalDrillDown) => {
        const originLocalIdentifier = getDrillOriginLocalIdentifier(globalDrillDown);
        const localIdentifier = `${originLocalIdentifier}_${objRefToString(globalDrillDown.target)}`;

        return {
            type: "attribute",
            drillTargetType: DRILL_TARGET_TYPE.DRILL_DOWN,
            attributeHierarchyRef: globalDrillDown.target,
            originLocalIdentifier: originLocalIdentifier,
            title: getTitleFromDrillableItemPushData(supportedItemsForWidget, originLocalIdentifier),
            attributes: [],
            localIdentifier,
            complete: true,
        };
    });
};
