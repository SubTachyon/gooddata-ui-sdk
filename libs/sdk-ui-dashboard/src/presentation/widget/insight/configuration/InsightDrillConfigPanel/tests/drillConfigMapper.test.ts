// (C) 2023 GoodData Corporation
import { describe, it, expect } from "vitest";

import { getGlobalDrillDownMappedConfigForWidget } from "../drillConfigMapper.js";
import { availableDrillTargets, globalDrillDowns } from "./drillConfigMapper.fixture.js";

describe("drillConfigMapper", () => {
    describe("getGlobalDrillDownMappedConfigForWidget", () => {
        it("should map global drill downs to drill down attribute hierarchy configs", () => {
            expect(
                getGlobalDrillDownMappedConfigForWidget(globalDrillDowns, availableDrillTargets),
            ).toMatchSnapshot();
        });
    });
});
