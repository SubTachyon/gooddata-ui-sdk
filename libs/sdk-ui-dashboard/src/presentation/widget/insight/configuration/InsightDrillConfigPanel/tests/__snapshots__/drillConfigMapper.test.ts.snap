// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`drillConfigMapper > getGlobalDrillDownMappedConfigForWidget > should map global drill downs to drill down attribute hierarchy configs 1`] = `
[
  {
    "attributeHierarchyRef": {
      "identifier": "477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_2",
      "type": "attributeHierarchy",
    },
    "attributes": [],
    "complete": true,
    "drillTargetType": "DrillDown",
    "localIdentifier": "458cb8556b604396be5d736ada2aa833_477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_2",
    "originLocalIdentifier": "458cb8556b604396be5d736ada2aa833",
    "title": "Region",
    "type": "attribute",
  },
  {
    "attributeHierarchyRef": {
      "identifier": "477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_2",
      "type": "attributeHierarchy",
    },
    "attributes": [],
    "complete": true,
    "drillTargetType": "DrillDown",
    "localIdentifier": "f20e093c6af5439dbfce10ee37620fc8_477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_2",
    "originLocalIdentifier": "f20e093c6af5439dbfce10ee37620fc8",
    "title": "Department",
    "type": "attribute",
  },
  {
    "attributeHierarchyRef": {
      "identifier": "477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_1",
      "type": "attributeHierarchy",
    },
    "attributes": [],
    "complete": true,
    "drillTargetType": "DrillDown",
    "localIdentifier": "458cb8556b604396be5d736ada2aa833_477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_1",
    "originLocalIdentifier": "458cb8556b604396be5d736ada2aa833",
    "title": "Region",
    "type": "attribute",
  },
  {
    "attributeHierarchyRef": {
      "identifier": "477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_1",
      "type": "attributeHierarchy",
    },
    "attributes": [],
    "complete": true,
    "drillTargetType": "DrillDown",
    "localIdentifier": "f20e093c6af5439dbfce10ee37620fc8_477f101d580549049144fdb8aa39f7b4_attribute_hierarchies_1",
    "originLocalIdentifier": "f20e093c6af5439dbfce10ee37620fc8",
    "title": "Department",
    "type": "attribute",
  },
]
`;
